lib/ansible/modules/network/frr/frr_facts.py

```diff
@@ -0,0 +1,470 @@
+EXAMPLES = """
+# Collect all facts from the device
+- frr_facts:
+    gather_subset: all
+
+# Collect only the config and default facts
+- frr_facts:
+    gather_subset:
+      - config
+
+# Do not collect hardware facts
+- frr_facts:
+    gather_subset:
+      - "!hardware"
+"""
+
+RETURN = """
+ansible_net_gather_subset:
+  description: The list of fact subsets collected from the device
+  returned: always
+  type: list
+
+# default
+ansible_net_hostname:
+  description: The configured hostname of the device
+  returned: always
+  type: str
+ansible_net_version:
+  description: The operating system version running on the remote device
+  returned: always
+  type: str
+
+# hardware
+ansible_net_mem_stats:
+  description: The memory statistics fetched from the device
+  returned: when hardware is configured
+  type: dict
+
+# config
+ansible_net_config:
+  description: The current active config from the device
+  returned: when config is configured
+  type: str
+
+# interfaces
+ansible_net_all_ipv4_addresses:
+  description: All IPv4 addresses configured on the device
+  returned: when interfaces is configured
+  type: list
+ansible_net_all_ipv6_addresses:
+  description: All IPv6 addresses configured on the device
+  returned: when interfaces is configured
+  type: list
+ansible_net_interfaces:
+  description: A hash of all interfaces running on the system
+  returned: when interfaces is configured
+  type: dict
+ansible_net_mpls_ldp_neighbors:
+  description: The list of MPLS LDP neighbors from the remote device
+  returned: when interfaces is configured and LDP daemon is running on the device
+  type: dict
+"""
+
+import re
+
+from ansible.module_utils.network.frr.frr import run_commands, get_capabilities
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.six import iteritems
+
+
+class FactsBase(object):
+
+    COMMANDS = list()
+
+    def __init__(self, module):
+        self.module = module
+        self.facts = dict()
+        self.responses = None
+        self._capabilities = get_capabilities(self.module)
+
+    def populate(self):
+        self.responses = run_commands(self.module, commands=self.COMMANDS, check_rc=False)
+
+    def run(self, cmd):
+        return run_commands(commands=cmd, check_rc=False)
+
+
+class Default(FactsBase):
+
+    COMMANDS = ['show version']
+
+    def populate(self):
+        super(Default, self).populate()
+        self.facts.update(self.platform_facts())
+
+    def platform_facts(self):
+        platform_facts = {}
+
+        resp = self._capabilities
+        device_info = resp['device_info']
+
+        platform_facts['system'] = device_info['network_os']
+
+        for item in ('version', 'hostname'):
+            val = device_info.get('network_os_%s' % item)
+            if val:
+                platform_facts[item] = val
+
+        return platform_facts
+
+
+class Hardware(FactsBase):
+
+    COMMANDS = ['show memory']
+
+    def _parse_total_heap_allocated(self, data):
```

Comment:
All these methods can be clubbed into a single method that takes data and pattern string as arguments and returns the match else None